### PR TITLE
[Object Detection] Updates to FasterRCNN

### PR DESCRIPTION
**Pull request by ianstenbit**
Main changes are:

- Make FasterRCNN a subclass of BaseObjectDetectionModel
- Add prediction decoding to the output of FasterRCNN.predict (required some changes to the NmsDecoder)
- Add TPU support for evaluation (including decoding)

Here's a [colab](https://colab.sandbox.google.com/drive/10hNe093WPyVx2wfFJD7AVzlvgwu2rh0-?resourcekey=0-Pa59dGS-88kQc3YspTOoUw#scrollTo=4sPJM_SMyChC) of FasterRCNN on TPU training against VOC/2007 and using our coco metrics callback. Resulting TensorBoard history pending

### CODE REVIEW

**Review by tanzhenyu — COMMENTED** [review]
Thanks for the PR!

File: keras_cv/models/object_detection/faster_rcnn.py [hunk 1]
```diff
@@ -226,7 +231,7 @@ def get_config(self):
 
 
 # TODO(tanzheny): add more configurations
-class FasterRCNN(tf.keras.Model):
+class FasterRCNN(ObjectDetectionBaseModel):
```
[the remark this answers is not in the document]
Comment (ianstenbit) [reply keras_cv/models/object_detection/faster_rcnn.py:234]: It gives us the in-graph CPU decoding logic in the override of `make_predict_function`.

I also think it's a good step toward standardizing the APIs across OD models to make them all subclass this. wdyt?

File: keras_cv/layers/object_detection/nms_prediction_decoder.py [hunk 2]
```diff
@@ -89,28 +89,33 @@ def _decode_box_predictions(self, anchor_boxes, box_predictions):
...
+        if anchor_boxes is None:
+            anchor_boxes = self.anchor_generator(images[0])
+            anchor_boxes = tf.concat(list(anchor_boxes.values()), axis=0)
+            anchor_boxes = bounding_box.convert_format(
+                anchor_boxes,
+                source=self.anchor_generator.bounding_box_format,
+                target="xywh",
+                images=images[0],
+            )
+            anchor_boxes = anchor_boxes[None, ...]
```
[the remark this answers is not in the document]
Comment (ianstenbit) [reply keras_cv/layers/object_detection/nms_prediction_decoder.py:118]: This is functionally equivalent to
```
anchor_boxes = tf.expand_dims(anchor_boxes, axis=0)
```
I just copied it as-is because this is how it was before. It's required because box_predictions has a batch axis, so anchor boxes needs one too

File: keras_cv/models/object_detection/faster_rcnn.py [hunk 3]
```diff
@@ -365,16 +383,12 @@ def call(self, images, training=None):
...
-                box_format=self.bounding_box_format,
-                variance=[0.1, 0.1, 0.2, 0.2],
-            )
-        return box_pred, cls_pred
+
+        if training:
+            return box_pred, cls_pred
+
+        # Return anchor boxes of interest during evaluation
+        return box_pred, cls_pred, rois
```
[the remark this answers is not in the document]
Comment (ianstenbit) [reply keras_cv/models/object_detection/faster_rcnn.py:394]: Good catch -- I think these need to be reformatted. This explains the bad scores I saw on my test run I think

[the remark this answers is not in the document]
Comment (ianstenbit) [reply keras_cv/models/object_detection/faster_rcnn.py:391]: Actually -- maybe these don't need to be reformatted.

It looks like the current FasterRCNN implementation doesn't actually support arbitrary bounding box format, it's got "yxyx" hardcoded. I'll need to fix that. I'll take a look tomorrow.

File: keras_cv/models/object_detection/faster_rcnn.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (tanzhenyu) [reply keras_cv/models/object_detection/faster_rcnn.py:234]: While it's nice to re-use logic through subclassing and override, it also somewhat defeats the purpose of the repo, which is to provide fork-able model code as a product, and less indirect abstractions.

For example, do we know if all OD models should inherit from this, or just anchor based models that need NMS?

If it's possible, I'd prefer not inherit from base model, and also remove base model from our repo.

File: keras_cv/models/object_detection/faster_rcnn.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (tanzhenyu) [reply keras_cv/models/object_detection/faster_rcnn.py:391]: I thought the `bounding_box_format` means the output format?

File: keras_cv/models/object_detection/faster_rcnn.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (LukeWood) [reply keras_cv/models/object_detection/faster_rcnn.py:234]: > If it's possible, I'd prefer not inherit from base model, and also remove base model from our repo.

Good point on the readability, I'm very torn on this.  On one hand, you're correct that the indirection is hard to follow.

On the other hand, OD base model has a LOT of code in it, and a LOT of logic in it; none related to modelling.

Maybe we can upstream keras APIs for asynchronous CPU decoding in the tf.data pipeline, and some sort of post-processing step that runs on CPU before metrics as Keras APIs.

File: keras_cv/models/object_detection/faster_rcnn.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (LukeWood) [reply keras_cv/models/object_detection/faster_rcnn.py:391]: it should represent input and output format both

File: keras_cv/models/object_detection/faster_rcnn.py [hunk 4]
```diff
@@ -286,10 +295,13 @@ def __init__(
         anchor_generator=None,
         rpn_head=None,
         rcnn_head=None,
+        prediction_decoder=None,
```
Comment (LukeWood) [inline keras_cv/models/object_detection/faster_rcnn.py:298]: can we default `include_rescaling` to None and make it required?  The reason we do this in `keras_cv.models` is to force users to think about how they preprocess their inputs.  Its a huge opportunity for silent failure to assume that they don't want rescaling (or to assume that they do want rescaling!)

Comment (ianstenbit) [reply keras_cv/models/object_detection/faster_rcnn.py:298]: I know there's been some disagreement about this, but I think for now we should at least make this consistent with the rest of the package, so this sgtm. I'm updating this to make include_rescaling required

Comment (martin-gorner) [reply keras_cv/models/object_detection/faster_rcnn.py:298]: +1 against a default value here (if that's what you meant).

File: keras_cv/models/object_detection/faster_rcnn.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ianstenbit) [reply keras_cv/models/object_detection/faster_rcnn.py:234]: I tend to agree that if possible we should avoid having the OD base model.

Right now (until we have NMS decoding working on TPU without the infra of the OD base model), this would result in a lot of duplicate code.

We could consider a refactor to try to use composition instead of inheritance, but that could get ugly since the shared logic is an override of base class methods.

[the remark this answers is not in the document]
Comment (martin-gorner) [reply keras_cv/models/object_detection/faster_rcnn.py:234]: +1 for keeping ODBaseModel as an API testing ground for now, then upstreaming common behaviors to Keras core (if possible and if desirable) and then removing ODBaseModel.
Shipping models with tons of duplicating code will not help readability. For people forking the models, in the short term, they will have to fork ODBaseModel too.

File: keras_cv/models/object_detection/faster_rcnn.py (hunk 4 above)
Comment (LukeWood) [reply keras_cv/models/object_detection/faster_rcnn.py:298]: > I know there's been some disagreement about this, but I think for now we should at least make this consistent with the rest of the package, so this sgtm. I'm updating this to make include_rescaling required

If theres disagreement we should discuss.

@fchollet @qlzh727 @martin-gorner and I discussed this thoroughly around a year ago and came to a strong consensus to fight the silent error.  Happy to discuss, but I feel confident in the required API.

File: keras_cv/models/object_detection/faster_rcnn.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (tanzhenyu) [reply keras_cv/models/object_detection/faster_rcnn.py:234]: What duplicate code are exactly required here, if we choose not to inherit?

File: keras_cv/models/object_detection/faster_rcnn.py (hunk 4 above)
Comment (ianstenbit) [reply keras_cv/models/object_detection/faster_rcnn.py:298]: sgtm. I agree with this decision, I just seem to recall some pushback about it (but it seems like I am likely misremembering!)

File: keras_cv/models/object_detection/faster_rcnn.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ianstenbit) [reply keras_cv/models/object_detection/faster_rcnn.py:234]: The big thing is `make_predict_function`, but we probably can/should eventually upstream more shared logic (e.g. a default NMS decoder?)

[the remark this answers is not in the document]
Comment (tanzhenyu) [reply keras_cv/models/object_detection/faster_rcnn.py:234]: IMHO, it's a tradeoff -- we only provide base classes when pros outweighs cons. The prediction / decoding is not a strong selling point 1) compared to asking users to subclass ODBaseModel when they implement YOLO3, so I wonder if we should resist the tempt to do so here, 2) for users that read our code.

At the end of the day, it's an advantage when we can merge common logic which is hard to deal with -- for example, export saved model for TFLite, or to support mixed precision, which is a difficult task and many users don't want to deal with it. But try to avoid another indirect layer seems more reasonable to me.

I will leave this up to you.

File: keras_cv/layers/object_detection/nms_prediction_decoder.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (tanzhenyu) [reply keras_cv/layers/object_detection/nms_prediction_decoder.py:118]: hmmm...can we use anchor_boxes[tf.newaxis, ...] instead then?

**Review by tanzhenyu — COMMENTED** [review]
comments inline

File: keras_cv/layers/object_detection/nms_prediction_decoder.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ianstenbit) [reply keras_cv/layers/object_detection/nms_prediction_decoder.py:118]: yeah that seems much more readable -- done

File: keras_cv/models/object_detection/faster_rcnn.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ianstenbit) [reply keras_cv/models/object_detection/faster_rcnn.py:234]: I think you're right that these should be split in the long-run. For now, I think the `make_predict_function` code is big enough and sensitive enough that it justifies avoiding duplicating the code. 

I'm going to put a TODO for now to eventually scrap the ODBaseModel once we feel like we have a stable API and (hopefully) once the TPU compat issues are solved upstream.

File: keras_cv/models/object_detection/faster_rcnn.py [hunk 5]
```diff
@@ -275,21 +280,32 @@ class FasterRCNN(tf.keras.Model):
             (all foreground classes + one background class). By default it uses the rcnn head
             from paper, which is 2 FC layer with 1024 dimension, 1 box regressor and 1
             softmax classifier.
+        prediction_decoder: (Optional)  A `keras.layer` that is responsible for
+            transforming RetinaNet predictions into usable bounding box Tensors.  If
```
Comment (tanzhenyu) [inline keras_cv/models/object_detection/faster_rcnn.py:284]: maybe we should clarify what the model output is before the decoder layer. "transforming RetinaNet predictions" sounds weird in this file.

File: examples/models/object_detection/faster_rcnn/train.py [hunk 6]
```diff
@@ -63,7 +63,9 @@
 eval_ds = tfds.load("voc/2007", split="test", with_info=False)
 
 with strategy.scope():
-    model = keras_cv.models.FasterRCNN(classes=20, bounding_box_format="yxyx")
+    model = keras_cv.models.FasterRCNN(
+        classes=20, bounding_box_format="yxyx", include_rescaling=True
```
Comment (tanzhenyu) [inline examples/models/object_detection/faster_rcnn/train.py:67]: I thought the training script uses `tf.keras.applications.resnet_50.preprocess_input`?

File: keras_cv/models/object_detection/faster_rcnn.py [hunk 7]
```diff
@@ -365,16 +387,12 @@ def call(self, images, training=None):
...
-                box_format=self.bounding_box_format,
-                variance=[0.1, 0.1, 0.2, 0.2],
-            )
-        return box_pred, cls_pred
+
+        if training:
+            return box_pred, cls_pred
+
+        # Return anchor boxes of interest during evaluation
+        return box_pred, cls_pred, rois
```
Comment (tanzhenyu) [inline keras_cv/models/object_detection/faster_rcnn.py:395]: any reason that we don't need `_decode_delta_to_boxes` anymore?

File: keras_cv/models/object_detection/faster_rcnn.py [hunk 8]
```diff
@@ -275,21 +280,32 @@ class FasterRCNN(tf.keras.Model):
...
+        include_rescaling,
         backbone=None,
-        include_rescaling=False,
         anchor_generator=None,
         rpn_head=None,
         rcnn_head=None,
+        prediction_decoder=None,
         **kwargs,
     ):
+        if bounding_box_format != "yxyx":
```
Comment (tanzhenyu) [inline keras_cv/models/object_detection/faster_rcnn.py:301]: any reason we add this constraint?

File: keras_cv/models/object_detection/faster_rcnn.py (hunk 5 above)
Comment (ianstenbit) [reply keras_cv/models/object_detection/faster_rcnn.py:284]: Good catch -- swapped RetinaNet for FasterRCNN and added explanation of box output before decoding.

File: examples/models/object_detection/faster_rcnn/train.py (hunk 6 above)
Comment (ianstenbit) [reply examples/models/object_detection/faster_rcnn/train.py:67]: Yes -- switched this to `False`

File: keras_cv/models/object_detection/faster_rcnn.py (hunk 7 above)
Comment (ianstenbit) [reply keras_cv/models/object_detection/faster_rcnn.py:395]: Decoding now happens in the `prediction_decoder`, so we don't need to call it here.

File: keras_cv/models/object_detection/faster_rcnn.py (hunk 8 above)
Comment (ianstenbit) [reply keras_cv/models/object_detection/faster_rcnn.py:301]: "yxyx" is hardcoded throughout as the box format. In a follow-up I plan to fix this, but for now I'm just raising an error because if another box format is used this will not behave correctly.

File: keras_cv/layers/object_detection/nms_prediction_decoder.py [hunk 9]
```diff
@@ -89,28 +89,33 @@ def _decode_box_predictions(self, anchor_boxes, box_predictions):
         )
         return boxes
 
-    def call(self, images, predictions):
+    def call(self, images, predictions, anchor_boxes=None):
```
Comment (tanzhenyu) [inline keras_cv/layers/object_detection/nms_prediction_decoder.py:92]: one more comment here: isn't the prediction a dict of boxes, classes and scores? or do we re-concatenate it somewhere else in the model?

**Comment by tanzhenyu** [comment]
This is being fixed in https://github.com/keras-team/keras-cv/pull/1061

**Comment by ianstenbit** [comment]
> This is being fixed in #1061

Ack -- closing this in favor of yours.
